### PR TITLE
[mlir][gpu] Add subgroup_reduce to shuffle lowering

### PR DESCRIPTION
**Pull request by kuhar**
This supports both the scalar and the vector multi-reduction cases.

### CODE REVIEW

**Comment by raikonenfnu** [comment]
This looks very similar to the IREE VectorReduction to shuffle lowering? Is the plan to lower vector.multi_reduce to gpu.subgroup_reduce and use this lowering instead?

**Comment by kuhar** [comment]
> This looks very similar to the IREE VectorReduction to shuffle lowering? Is the plan to lower vector.multi_reduce to gpu.subgroup_reduce and use this lowering instead?

Yes, exactly. I described this lowering flow here: https://github.com/llvm/llvm-project/pull/76015

**Comment by raikonenfnu** [comment]
> #76015

Woaahhhh, exciting! :)

File: mlir/lib/Dialect/GPU/Transforms/SubgroupReduceLowering.cpp [hunk 1]
```diff
@@ -139,6 +143,164 @@ struct ScalarizeSingleElementReduce final
   }
 };
 
+/// Emits a subgroup reduction using a sequence of shuffles. Uses the `packFn`
+/// and `unpackFn` to convert to/from the native shuffle type. Assumes that the
```
Comment (Groverkss) [inline mlir/lib/Dialect/GPU/Transforms/SubgroupReduceLowering.cpp:147]: Can you write more here on what is expected from `pack` and `unpackFn`?

File: mlir/test/lib/Dialect/GPU/TestGpuRewrite.cpp [hunk 2]
```diff
@@ -47,19 +48,37 @@ struct TestGpuSubgroupReduceLoweringPass
...
+      *this, "expand-to-shuffles",
+      llvm::cl::desc("Expand subgroup_reduce ops to shuffle ops."),
+      llvm::cl::init(false)};
+
   void runOnOperation() override {
     RewritePatternSet patterns(&getContext());
     populateGpuBreakDownSubgrupReducePatterns(patterns,
-                                              /*maxShuffleBitwidth=*/32);
+                                              /*maxShuffleBitwidth=*/32,
+                                              PatternBenefit(2));
```
Comment (Groverkss) [inline mlir/test/lib/Dialect/GPU/TestGpuRewrite.cpp:77]: Can you add a comment here on why you set the benefit to a non default value?

**Review by Groverkss — APPROVED** [review]
Overall LGTM. I just have some documentation related comments.

**Pull request merged**